mono/profiler/proflog.c

```diff
@@ -959,16 +1074,14 @@ remove_thread (MonoProfiler *prof, MonoProfilerThread *thread, gboolean from_cal
 
 			uint64_t now = current_time ();
 
-			ENTER_LOG (buffer, "thread-end");
```

Comment:
It was there to catch things like:

```c
static void
some_callback (MonoProfiler *prof)
{
    ENTER_LOG;

    LogBuffer *buf = ensure_logbuf (...);

    /* write some data to buf */

    /* get some data from the runtime, triggers some_other_callback */
    stuff = get_something_from_runtime ();

    /* write stuff to buf */

    EXIT_LOG;
}

static void
some_other_callback (MonoProfiler *prof)
{
    ENTER_LOG;

    /* here we'd end up writing in the middle of another event */
}
```

The correct way to write the above would be to call `get_something_from_runtime` outside of the `ENTER_LOG`/`EXIT_LOG` pair.

I've never seen this catch any actual errors in the profiler, though, and I've gone through and checked all event callbacks to make sure they don't call runtime functions while writing to a buffer. Unfortunately, even if this debugging code would have caught problems, all it would do is write to `stderr`, not assert or anything. So there's a good chance it would just be missed.

That said, the reason I removed this debugging code was that I originally intended to remove the `ENTER_LOG`/`EXIT_LOG` macros. But as you can see, I settled on keeping them to create a scope around buffer writing. So, we could just replace the old debugging code with some code that actually asserts if such a case is ever hit.